File: docs/research/corpus-examples/auth-session/python/vulnerable-4-altered.py
"""Manufactured third variant for auth-session/python CWE-330/CWE-338
(Use of Insufficiently Random Values / Cryptographically Weak PRNG) --
distinct from this cell's natural pair (vulnerable-2.py/idiomatic-2.py,
DOAJ's real ``random.randint``-based login code). This one targets a
different token purpose and a different framework idiom: a Django-style
custom auth backend's "remember me" cookie token, built by hashing a
``random.random()`` draw instead of drawing directly from a CSPRNG.

Manufactured per docs/VULN_CORPUS_PAIR_MANUFACTURING_PLAN.md's "Per-pair
mechanics".
"""
from __future__ import annotations

import hashlib
import random


def generate_remember_me_token(user_id: int) -> str:
    """BUG: ``random`` is CPython's Mersenne Twister (MT19937), not a
    CSPRNG -- its internal state is recoverable from a handful of
    observed outputs. Hashing a low-entropy ``random.random()`` draw with
    SHA-256 does not add real entropy; it only obscures a still-
    predictable input."""
    raw = f"{user_id}-{random.random()}"
    return hashlib.sha256(raw.encode()).hexdigest()

File: docs/research/corpus-examples/auth-session/node/vulnerable-4-altered.js
// Manufactured third variant for auth-session/node CWE-330/CWE-338 (Use
// of Insufficiently Random Values / Cryptographically Weak PRNG) --
// distinct from this cell's natural pair (vulnerable-2.js/idiomatic-2.js,
// which is about the express-session HMAC *secret*). This one targets a
// different token purpose entirely: a password-reset token generator, a
// common Express/Koa-style route handler idiom.
'use strict';

// BUG: builds the reset token out of Date.now() (a coarse, often-logged/
// guessable timestamp) and Math.random() (a non-cryptographic PRNG whose
// internal state is recoverable from a handful of observed outputs) --
// neither component is suitable as a security-sensitive secret.
function generateResetToken() {
  return Date.now().toString(36) + Math.random().toString(36).slice(2);
}

module.exports = { generateResetToken };
